src/librustdoc/clean/utils.rs

```diff
@@ -468,7 +468,7 @@ pub(crate) fn resolve_type(cx: &mut DocContext<'_>, path: Path) -> Type {
     match path.res {
         Res::PrimTy(p) => Primitive(PrimitiveType::from(p)),
         Res::SelfTyParam { .. } | Res::SelfTyAlias { .. } if path.segments.len() == 1 => {
```

Comment:
I don't believe that the handling of `SelfTyAlias` is correct here 🤔 Or at least I do want to point out that `SelfTyParam` and `SelfTyAlias` act differently and should probably not continue to be conflated.

`SelfTyAlias` is used for the self type of an impl; this is not some abstract `SelfTy` (which is literally just parameter -- unrelated, but I'm not sure why this PR moves away from representing self as a parameter when it is one in `trait`, but I digress) -- `SelfTyAlias` is a specific, sometimes concrete type. For example, in:

```rust
impl MyCustomType<i32> { /* Self is a SelfTypeAlias here */ }
```

Reply:
Yes, thanks for raising this. I believe that some other part of rustdoc is "unwrapping" the `Self` alias; see for example [`Vec::new()`](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.new)'s return type. It may make sense to centralize that behavior in one place—namely here—though.

Regarding your comment about the change in representation: I understand that `Self` in trait is logically a parameter (à la Haskell type classes) in the compiler's view. However, from the perspective of users and many of Rustdoc's transformations, we treat it specially and don't think of it as such. For example, we don't normally render it as a parameter on a trait's page (though we do currently have some bugs, like the `: PartialEq<Self>` [here](https://doc.rust-lang.org/nightly/core/cmp/trait.Eq.html) but not [here](https://doc.rust-lang.org/nightly/std/cmp/trait.Eq.html)). So I feel like from Rustdoc's perspective, it makes sense to treat it specially unless and until we replace `clean::Type` with `rustc_middle::ty::Ty`. Thoughts?

I do intend to add information about the meaning of the type (alias, with value, or parameter, with `DefId`) to `Type::SelfTy`, but rustdoc currently doesn't use that information.

Reply:
> Yes, thanks for raising this. I believe that some other part of rustdoc is "unwrapping" the `Self` alias; see for example [`Vec::new()`](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.new)'s return type.

No, that's not the reason. You've linked to `std`'s `Vec::new` which is an inlined cross-crate re-export. If you take a look at the local docs in `alloc`: https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html#method.new, you can see that rustdoc retains the literal `Self`.

This is due to the fact that in the local case rustdoc cleans the HIR which contains the still-unresolved `Self` path. In the IXCRE case, rustdoc cleans rustc_middle::ty types were the SelfTyAlias has been resolved to a 'concrete' type (here: `Vec<T>`). In rustc_middle::ty only SelfTyParam is retained (as `tcx.types.self_param` / "`TyKind::Param(0, kw::SelfUpper)`").

This is bug (or discrepancy at least) is tracked in #44306. See my latest comment from almost a year ago: https://github.com/rust-lang/rust/issues/44306#issuecomment-1785819041 describing how we might go about fixing it and why I haven't taken it on (yet). Note that that comment assumes we want to render `Self` over the resolved aliased type which might lead to better UX (unknown).